src/main/java/com/capitalone/dashboard/collector/DefaultGitlabClient.java

```diff
@@ -192,28 +227,29 @@ public Build getPipelineDetails(String buildUrl, String instanceUrl) {
             JSONParser parser = new JSONParser();
             try {
                 JSONObject buildJson = (JSONObject) parser.parse(resultJSON);
-                Boolean building = "running".equalsIgnoreCase(getString(buildJson, "status"));
-                Boolean cancelled = "canceled".equalsIgnoreCase(getString(buildJson, "status"));
-
-                if (!building && !cancelled) {
-                    Build build = new Build();
-                    build.setNumber(buildJson.get("id").toString());
-                    build.setBuildUrl(buildUrl);
-                    build.setTimestamp(jobsForPipeline.getEarliestStartTime());
-                    build.setStartTime(jobsForPipeline.getEarliestStartTime());
-                    build.setDuration(jobsForPipeline.getRelevantJobTime());
-                    build.setEndTime(jobsForPipeline.getLastEndTime());
-                    build.setBuildStatus(getBuildStatus(buildJson));
+
+                if (isBuildCompleted(buildJson) && jobsForPipeline != null) {
+                    Build build = getBuild(buildUrl, jobsForPipeline, buildJson);
                     Iterable<String> commitIds = jobsForPipeline.getCommitIds();
                     List<Commit> commits = new ArrayList<>();
                     if (commitIds != null) {
                         commitIds.forEach(commitId -> {
-                            List<Commit> revisionNumbers = commitRepository.findByScmRevisionNumber(commitId).stream()
-                                    .filter(c -> c.getScmParentRevisionNumbers().size() > 1) /* Extract only merge commits */
-                                    .collect(Collectors.toList());
-                            if (revisionNumbers != null && !revisionNumbers.isEmpty()) {
-                                commits.addAll(revisionNumbers);
+                            List<Commit> matchedCommits = commitRepository.findByScmRevisionNumber(commitId);
+                            Commit newCommit;
+                            if (matchedCommits != null && matchedCommits.size() > 0) {
+                                newCommit = matchedCommits.get(0);
+                            } else {
+                                newCommit = getCommit(commitId, instanceUrl, gitProjectId);
+
+                            }
+
```

Comment:
Can you also extract out this piece of code?
I think we can also write it this way, reducing the number of lines -
```suggestion
                               Commit newCommit = commitRepository.findByScmRevisionNumber(commitId).stream()
                                    .findFirst()
                                    .orElse(getCommit(commitId, instanceUrl, gitProjectId));
```

Reply:
I also think that `getCommit` should return Optional rather than null so that the contract clearly tells that there is a chance of no element being returned here. I can discuss with you in person regarding this.